src/GoTrueClient.ts

```diff
@@ -471,14 +474,18 @@ export default class GoTrueClient {
       if (data.session) {
         await this._saveSession(data.session)
         await this._notifyAllSubscribers('SIGNED_IN', session)
+      } else if (currentSession) {
+        await this._saveSession(currentSession)
       }
 
       return { data: { user, session }, error: null }
     } catch (error) {
+      if (currentSession) {
+        await this._saveSession(currentSession)
+      }
```

Comment:
i'm not sure why it's necessary to preserve the existing session when one calls the signup method - we always want to remove the session when signup is called to prevent any possibilities of the user logging in as someone else.

Reply:
this applies to all the other sign-in methods too - if you're signing in, there won't be any session in the first place, else you'd be logged in already

Reply:
This truly depends on the UI implementation. We may want to allow users to directly switch accounts, so it is totally possible to call sign in from being already signed in form another account. Right now, if something fails, the user is being kicked out of their current session. This applies to resend events like email change that can also happen while logged in.

I also dont see how this would log a user to someone else's account. Either the signin is successful and the current session switches to the new one, or it fails and you keep the existing one.

Reply:
If the team wanted to go down this path, you could simplify a lot of these methods by moving the _removeSession call into the `if` blocks which check for `data.session` or `session?.access_token`. There would be a few exceptions to that though.

I've seen more and more discussions about people trying to implement a "switch account" feature; or they already have, but recent changes have broken their implementations.

Reply:
I agree. I just wasn't sure if some logic was dependent on having no session while executing, so I preferred keeping it and reverting at the end.

Hopefully, we can remove the current limitation, be it by removing `_removeSession` where not necessary or, as my PR suggests, restoring the session on fail.